src/ckanext-natcap/ckanext/natcap/plugin.py

```diff
@@ -240,6 +243,121 @@ def natcap_update_mappreview(context, package):
     #NatcapPlugin._after_dataset_update(context, package_data)
 
 
+def _load_download_rules_for(pkg):
+    """Try to get dataset-specific rules (first by title, then name)"""
+    rules_dir = path.join(
+        path.dirname(__file__), 'dataset_configs'
+    )
+    if not path.isdir(rules_dir):
+        LOGGER.debug(f"dataset_configs not found at: {rules_dir}")
+
+    candidates = []
+    if pkg:
+        if pkg.get('title'):
+            candidates.append(path.join(rules_dir, f"{pkg['title']}.yml"))
+        if pkg.get('name'):
+            candidates.append(path.join(rules_dir, f"{pkg['name']}.yml"))
+            # name will be like: sts-12a3b4567 etc.
+
+    for fp in candidates:
+        try:
+            if path.exists(fp):
+                with open(fp, 'r') as f:
+                    data = yaml.safe_load(f) or {}
+                    LOGGER.info(f"Loaded per-dataset rules: {fp}")
+                    return data
+        except Exception:
+            LOGGER.exception(f"Failed loading per-dataset rules from {fp}")
+
+    return {}
+
+
+def _apply_rule_list(rule_list, relpath, base, ext, url):
+    """Evaluate first-match-wins
+
+    Returns:
+        dict: {'allowed': bool, 'reason': str} or None if none match."""
+    for rule in rule_list:
+        m = (rule.get('match') or {})
+        if _match_rule(m, relpath, base, ext):
+            return {'allowed': bool(rule.get('allow', True)),
+                    'reason': rule.get('reason', ''),
+                    'url': url}
+    return None
+
+
+def _match_rule(m, relpath, base, ext):
+    pg = m.get('path_glob')
+    if pg and not fnmatch.fnmatch(relpath, pg):
+        return False
+    rx = m.get('name_regex')
+    if rx:
+        try:
+            if not re.search(rx, base):
+                return False
+        except re.error:
+            return False
+
+    ex = m.get('ext_any')
+    if ex and ext.lower() not in [e.lower() for e in ex]:
+        return False
+    return True
+
+
+def get_directory_url(node, target_name):
+    """Infer the URL of a zipfile of a directory"""
+    if node["type"] == "directory":
+        if node["name"] == target_name:
+            # infer URL from first child's URL
+            for child in node.get("children", []):
+                if "url" in child:
+                    # Drop the filename and change dir path to zip path
+                    return "/".join(child["url"].split("/")[:-1])+".zip"
+
+        # recurse into children
+        for child in node.get("children", []):
+            result = get_directory_url(child, target_name)
+            if result:
+                return result
+    return None
+
+
+def get_file_downloadability(pkg, source):
+    """
+    Check if a 'source' (file/dir in sources_list.html) is downloadable.
+
+    Returns:
+        dict: {'allowed': bool, 'reason': '...'}
+    """
```

Comment:
I think I'm inferring from the function itself that `source` is either a dict or an object(?) or maybe something else.  It would be really helpful to define and document the contract up front if we can and then stick to it!